tests/pallas/tpu_pallas_interpret_test.py

```diff
@@ -0,0 +1,974 @@
+import jax.numpy as jnp
+
+jax.config.parse_flags_with_absl()
+
+P = jax.sharding.PartitionSpec
+
+
+# Run all tests with 8 CPU devices.
+_exit_stack = contextlib.ExitStack()
+
+def setUpModule():
+  _exit_stack.enter_context(jtu.set_host_platform_device_count(8))
+
+def tearDownModule():
+  _exit_stack.close()
+
+class InterpretTest(jtu.JaxTestCase):
+
+  def test_right_permute_example(self):
+    num_devices = jax.device_count()
+    partition = P(None, 'x')
+    mesh = jax.make_mesh((num_devices,), ('x',))
+    sharding = jax.sharding.NamedSharding(mesh, partition)
+
+    # Create an input array that shards the last dimension across
+    # all devices.
+    input_arr = jax.random.uniform(
+      jax.random.key(0), (8, 128 * num_devices), dtype=jnp.float32)
+    input_arr = jax.device_put(input_arr, sharding)
+
+    def right_permute_kernel(input_ref, output_ref, send_sem, recv_sem):
+      my_id = lax.axis_index('x')
+      left_neighbor = lax.rem(my_id + num_devices - 1, jnp.int32(num_devices))
+      right_neighbor = lax.rem(my_id + 1, jnp.int32(num_devices))
+
+      barrier_sem = pltpu.get_barrier_semaphore()
+      def _body(ijk):
+        i, (j, k) = ijk
+        lax.cond(
+          (i == 0) | (j == 0),
+          lambda: pltpu.semaphore_signal(
+            barrier_sem,
+            device_id=(left_neighbor,),
+            device_id_type=pltpu.DeviceIdType.MESH),
+          lambda: pltpu.semaphore_signal(
+            barrier_sem,
+            device_id=(right_neighbor,),
+            device_id_type=pltpu.DeviceIdType.MESH))
+        return (i + 1, (j + 1, k + 1))
+      lax.while_loop(lambda ijk: ijk[0] < 2, _body, (0, (0, 0)))
+      pltpu.semaphore_wait(barrier_sem, 2)
+
+      def _body2(i, a):
+        remote_copy_op = pltpu.make_async_remote_copy(
+          src_ref=input_ref,
+          dst_ref=output_ref,
+          send_sem=send_sem,
+          recv_sem=recv_sem,
+          device_id=(right_neighbor,),
+          device_id_type=pltpu.DeviceIdType.MESH,
+        )
+        remote_copy_op.start()
+        remote_copy_op.wait()
+        return i + 1, a + 1
+      _ = lax.scan(_body2, 0, jnp.arange(4.0), unroll=2)
+
+    out_shape = jax.ShapeDtypeStruct((8, 128), jnp.float32)
+    grid_spec = pltpu.PrefetchScalarGridSpec(
+        num_scalar_prefetch=0,
+        # TPUMemorySpace.ANY will (usually) place the tensor in HBM.
+        in_specs=[
+            pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
+        ],
+        out_specs=pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
+        scratch_shapes=(
+            # We allocate DMA semaphores in scratch memory.
+            [pltpu.SemaphoreType.DMA] * 2
+        ),
+    )
+    right_permute = pl.pallas_call(
+        right_permute_kernel,
+        out_shape=out_shape,
+        grid_spec=grid_spec,
+        compiler_params=pltpu.TPUCompilerParams(collective_id=13),
+        interpret=True,
+        backend='mosaic_tpu',
```

Comment:
Idea: backend="cpu_like_mosaic_tpu"

Reply:
Yea that seems good. Or "tpu_verification" if the main purpose will be for verification.